### PR TITLE
Fixes #382 Changed Add Account to use account model name

### LINKED ISSUE #382: Add account form should output the type of account
In the flow of adding an account, by the time you get to the form it should say which account type you're adding.

So if you clicked "Bank account" it should say "Add bank account", for example.

This is located in `views/accounts/new.html.erb`

CleanShot 2024-02-08 at 12 17 07@2x

### PR DESCRIPTION
**Pull request by 2000rosser**
Fixes #382 The change currently only works for adding a 'Bank Account' as adding other types of accounts only creates a new Bank Account. When new types of accounts are added this will use the account name in the models as the title.

Signed-off-by: 2000rosser

### CODE REVIEW

**Comment by 2000rosser** [comment]
I believe this new change will resolve the vulnerability scan issue

File: app/views/accounts/new.html.erb [hunk 1]
```diff
@@ -65,11 +65,19 @@
...
+          <% allowable_types = ["Depository", "Investment", "Property", "Vehicle", "Credit", "Loan", "OtherAsset", "OtherLiability"] %>
+          <% if allowable_types.include?(params[:type]) %>
+            <% model_class = "Account::#{params[:type]}".constantize %>
+            Add <%= model_class.model_name.human(count: 2) %>
+          <% else %>
+            Add Account
+          <% end %>
+        <% else %>
+          Add Account
+        <% end %>
```
Comment (Shpigford) [inline app/views/accounts/new.html.erb:78]: Super heavy on the if/else statements, which feels a little hard to follow/maintain.

What about something like this? Not 100% positive it'll pass the ruby checks. 
```
<span>
        <% model_class = if params[:type].in?(["Depository", "Investment", "Property", "Vehicle", "Credit", "Loan", "OtherAsset", "OtherLiability"])
                  "Account::#{params[:type]}".constantize
                end %>
        Add <%= model_class ? model_class.model_name.human(count: 2).downcase : "account" %>
      </span>
```

Comment (2000rosser) [reply app/views/accounts/new.html.erb:78]: Yeah definitely worth a shot, I'll add it now

Comment (robzolkos) [reply app/views/accounts/new.html.erb:78]: We should centralize/codify the allowable types as I'm sure they will be queried everywhere

```ruby
  # in Account model
  def self.allowable_accountable_types
    accountable_types.map { |type| type.demodulize }
  end
```

Then can do 

```ruby
<% if params[:type].in? Account.allowable_account_types %>
 ... 
<% else %>
   Add Account
<% end %>
```

And probably best to move to a helper to clean up conditionals out of the view.

Comment (2000rosser) [reply app/views/accounts/new.html.erb]: Could it be worthwhile adding new models for singular and plural cases?
```yml
models:
      account/depository:
        singular: Bank Account
        plural: Bank Accounts
```

Comment (robzolkos) [reply app/views/accounts/new.html.erb]: Rails handles that with the `pluralize` methods, so I'm not sure we do (yet)

Comment (2000rosser) [reply app/views/accounts/new.html.erb]: The current implementation has Bank Accounts and Investments as plural by default, should they be changed to singular then?

Comment (2000rosser) [reply app/views/accounts/new.html.erb]: My bad 😆

**Comment by Shpigford** [comment]
Starting to wonder about how we're trying to be clever with the use of `constantize` and the account types.

It's causing problems in at least 3 different PRs right now.

Wondering if we should abandon ship on that and use something maybe slightly less succinct.

```
Confidence: High
Category: Remote Code Execution
Check: UnsafeReflection
Message: Unsafe reflection method `constantize` called on parameter value
Code: "Account::#{params[:type]}".constantize
File: app/views/accounts/new.html.erb
```

**Comment by Shpigford** [comment]
This was ultimately addressed in #395

**Pull request closed without merging**